[qa] Add logging to test_framework.py

jnewbery opened the pull request:
This is a WIP PR. Comments and feedback on the overall approach welcomed!

(For an overview of python's logging module and terms, see https://docs.python.org/3/library/logging.html and https://docs.python.org/3/howto/logging.html)

This PR adds logging to the test-framwork.py module. The "TestFramework" logger is added as a member of the BitcoinTestFramework class, and two handlers are attached to the logger:

- a StreamHandler() which outputs ERROR and higher level records to stderr.
- a FileHandler() which outputs all log records to a new `test_framework.log` file in the temporary test directory.

Test scripts and helper modules can log to the same Logger so there's a single log output:

- Individual test scripts can log to "TestFramework" by calling `self.log.[debug|info|warning|error|critical|exception]()`.
- Helper modules should log to child Loggers such as "TestFramework.mininode" or "TestFramework.authproxy", which will result in log records being propagated up to the "TestFramework" Logger.

I've added examples of how individual scripts and helper modules should log events in p2p-segwit.py and mininode.py.

The eventual goal is to remove all `print()` calls and replace them with logging.

A few notes on the implementation:

- a new `--loglevel` parameter can be used to log lower level records to the console, for example `--loglevel=DEBUG` will emit all log records to the console.
- the date format of `test_framework.log` is the same as bitcoind's debug.log. It should be trivial to interleave `test_framework.log` with the various `nodeX/debug.log` files for a global view of what's happening during the test.
- this should have no interaction with the existing `--tracerpc` parameter

jnewbery committed:
Add logging to test_framework.py

This commit adds python logging to test_framework.py. By default this
will output all log levels (DEBUG-INFO-WARNING-ERROR-CRITICAL) to a
test_framework.log file in the temporary test directory, and higher
level logs (WARNING-ERROR-CRITICAL) to the console. The level of logging
to the console can be controlled by a new log-level parameter.

This should have no interaction with the existing trace-rpc parameter.

jnewbery committed:
Add logging to p2p-segwit.py

jnewbery committed:
Use logging in mininode.py

This commit adds a TestFramework.mininode Logger to the mininode module.
This is a child logger of TestFramework, so the handlers set up in
test_framework.py will receive records from this logger and emit them
to the log file and console as appropriate.

laanwj commented:
If you're working on custom logging stuff anyway: what would be really nice is something that would queue up debug messages, then print them only if the test fails. This lowers the amount of spam in the Travis log in the good case, and gives full troubleshooting information in the bad case.

jnewbery commented:
@laanwj done: https://github.com/bitcoin/bitcoin/pull/9780

ryanofsky commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -183,12 +179,41 @@ def main(self):
+    def start_logging(self):
+        # Add logger and logging handlers
+        self.log = logging.getLogger('TestFramework')
+        self.log.setLevel(logging.DEBUG)
```
I think you're better off dropping this call. The default debug level seems to be 0, and the value of DEBUG is 10 so this potentially throws away logs.

Also would drop the 3 other setLevel(logging.DEBUG) calls below.

jnewbery replied:
I think DEBUG is good here. We shouldn't encourage people to log messages at custom levels (ie not one of debug, info, warning, error and critical), and we especially shouldn't encourage people to log at a level lower than DEBUG.

ryanofsky replied:
> I think DEBUG is good here. We shouldn't encourage people to log messages at custom levels (ie not one of debug, info, warning, error and critical), and we especially shouldn't encourage people to log at a level lower than DEBUG.

Actually setting DEBUG seems to be needed here in order to be able to see debug messages at all, otherwise python seems to fall back to root logger level which is WARNING.

Still I would consider dropping the "fh.setLevel()" call below. Having that call seems less like "not encouraging" people to log messages at custom levels, and more like going out of our way to punish and confuse them.

jnewbery replied:
I still don't agree. logging at non-standard levels is harder (you need to call `self.log.log(lvl, message`) rather than `self.log.debug(message)` or similar). Why would anyone want do that, and why would we want to allow people to add confusing logging levels to the test cases?

ryanofsky commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -183,12 +179,41 @@ def main(self):
+        fh.setLevel(logging.DEBUG)
+        # Create console handler to log messages to stderr. By default this logs only error messages, but can be configured with --loglevel.
+        ch = logging.StreamHandler()
+        console_log_level = getattr(logging, self.options.loglevel.upper(), None)
```
Calling getattr doesn't seem to be necessary, the documentation says setLevel already accepts a string: 
https://docs.python.org/3/library/logging.html#logging.Handler.setLevel. Maybe upper() is still useful though.

jnewbery replied:
Agree. I was following the Python 'logging how to' guide here: https://docs.python.org/3/howto/logging.html#logging-to-a-file which uses getattr(). I guess that was written before 3.2, which is when setlevel() was updated to take a string argument.

I've tested the logging module and:
- we do need to use upper() to set the log levels to uppercase
- calling with an invalid string causes logging to raise a helpful exception (I was worried that it might fail silently).

So calling setlevel directly with self.options.loglevel.upper() is better.

ryanofsky commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -183,12 +179,41 @@ def main(self):
+            rpc_handler = logging.StreamHandler()
+            rpc_handler.setLevel(logging.DEBUG)
+            rpc_logger.addHandler(rpc_handler)
+            rpc_logger.debug("debug")
```
Junk lines?

jnewbery replied:
yes. Junk lines from when I was testing this. Thanks. Removed.

ryanofsky commented:
3e7ed8c21c3f52a312a538746aebbbb06f67e3a1

jnewbery commented:
@ryanofsky thanks for the great feedback. I'm going to hold back on making any changes until I get some more concept ACKs/NACKs. I want to make sure people are happy going in this direction before I spend any time addressing review comments.

jnewbery commented:
@MarcoFalke @laanwj @jonasschnelli @sdaftuar you're the most active contributors to the qa directory. Can you have a quick look and give me a concept ACK/NACK so I know whether to continue in this direction.

No rush. Obviously you're all busy with the 0.14 release.

jnewbery committed:
Always enable debug log and microsecond logging for test nodes.

jnewbery committed:
Remove manual debug settings in qa tests.

-debug and -logtimemicros are now set by default. Individual test cases
no longer need to set these parameters manually.

jnewbery committed:
Use logging in test_framework/util.py

jnewbery committed:
Use logging in test_framework/blockstore.py

jnewbery committed:
Use logging in test_framework/comptool.py

sdaftuar commented:
Concept ACK.  

I haven't reviewed this PR in detail yet, but what I'd find useful is if the default behavior when running a single test locally was to print all log messages to the console, and only suppress log messages when running the tests through the `rpc-tests.py` script.

jnewbery commented:
@sdaftuar the eventual aim will be to add a lot more low-level logging (eg logging when every p2p message is sent or received on the mininode connection). Those low-level logs will always be saved to test_framework.log, but by default are not printed to the console. If the user want low-level logs printed to console, they can specify --loglevel when running the script.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -113,6 +113,8 @@ def main(self):
                           help="Directory for caching pregenerated datadirs")
         parser.add_option("--tmpdir", dest="tmpdir", default=tempfile.mkdtemp(prefix="test"),
                           help="Root directory for datadirs")
+        parser.add_option("--loglevel", dest="loglevel", default="ERROR",
```
Thanks. Updated.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -113,6 +113,8 @@ def main(self):
         parser.add_option("--tmpdir", dest="tmpdir", default=tempfile.mkdtemp(prefix="test"),
                           help="Root directory for datadirs")
+        parser.add_option("--loglevel", dest="loglevel", default="ERROR",
+                          help="log events at this level and higher to the console")
```
Updated

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -140,38 +139,35 @@ def main(self):
         except JSONRPCException as e:
-            print("JSONRPC error: "+e.error['message'])
-            traceback.print_tb(sys.exc_info()[2])
+            self.log.exception("JSONRPC error: "+e.error['message'])
```
Yep, good catch.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -140,38 +139,35 @@ def main(self):
                 os.rmdir(self.options.root)
         else:
-            print("Not cleaning up dir %s" % self.options.tmpdir)
+            self.log.info("Not cleaning up dir %s" % self.options.tmpdir)
```
good point. Fixed.

jnewbery commented:
I've addressed @ryanofsky's review comments and repushed. A few updates since I opened this PR:

- microsecond debug logging is now set by default on all test nodes. I've removed all places in the test code where debugging is set manually, since there's no reason not to use debug logging in these test cases.
- I've added logging to the test_framework submodules
- I've replaced print functions in the individual test-cases with logging calls

I've been running with this patch set locally for some time, together with a tool to interleave test_framwork logs with individual node logs. It makes troubleshooting testcases a lot easier since I have a global view of everything that's happening in the test case (ie I can see messages being sent in by the test framework and then received by the node).

I believe this is ready for merge, but any further review comments are welcomed.

ryanofsky reviewed (commented):
ACK 3b336aba4ef250868d6f3dc2bf741b0bbf621d39 assuming travis tests will be fixed.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -112,6 +112,8 @@ def main(self):
                           help="Directory for caching pregenerated datadirs")
         parser.add_option("--tmpdir", dest="tmpdir", default=tempfile.mkdtemp(prefix="test"),
                           help="Root directory for datadirs")
+        parser.add_option("--loglevel", dest="loglevel", default="ERROR",
```
You're right. This should be INFO by default. DEBUG emits way too much to the console to be the default.

pull-tester diverts stdout from the subtests, so there's no output to console at all.

--loglevel now accepts ints as well as the named strings and can be specified with the short option -l

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -139,38 +138,35 @@ def main(self):
         success = False
         try:
             os.makedirs(self.options.tmpdir, exist_ok=False)
```
Good idea. I needed to put start_logging() below the os.makedirs() call, but you're right that this should be outside the large try/except block. I've now moved them both up.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -182,12 +178,36 @@ def main(self):
             sys.exit(0)
         else:
-            print("Failed")
+            self.log.warning("Failed")
```
Agree. Done both.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -182,12 +178,36 @@ def main(self):
+        fh.setLevel(logging.DEBUG)
+        # Create console handler to log messages to stderr. By default this logs only error messages, but can be configured with --loglevel.
+        ch = logging.StreamHandler()
+        ch.setLevel(self.options.loglevel.upper())
```
Yep. Good idea. Done (although I've split this over two lines for readability).

jnewbery commented on qa/rpc-tests/bip68-sequence.py:
```diff
@@ -34,26 +34,26 @@ def run_test(self):
         self.activateCSV()
 
-        print("Verifying nVersion=2 transactions are standard.")
-        print("Note that with current versions of bitcoin software, nVersion=2 transactions are always standard (independent of BIP68 activation status).")
```
Because I previously thought that it was a bug that Core v0.13 (and a few earlier versions) always relayed v2 transactions - hence the 'with current versions' comment. See #9739 for the discussion. Whether or not it's a bug is moot now, and I think the 'with current verions' comment doesn't clarify.

'always' here means 'both before and after BIP 68 activation'.

jnewbery commented on qa/rpc-tests/bip68-sequence.py:
```diff
@@ -34,26 +34,26 @@ def run_test(self):
         self.test_version2_relay()
 
-        print("Passed\n")
+        self.log.info("Passed\n")
```
Thanks. Done.

jnewbery commented on qa/rpc-tests/test_framework/comptool.py:
```diff
@@ -21,6 +21,10 @@
 
+import logging
+
+logger=logging("TestFramework.comptool")
```
Yep. My mistake. Fixed.

jnewbery commented:
@ryanofsky's comments addressed and repushed: c9755da51a742c0d86572f750e16573e87fdd166 / https://github.com/jnewbery/bitcoin/tree/pr9768.3

jnewbery commented:
rebased: https://github.com/bitcoin/bitcoin/pull/9768/commits/2c9fc0e5f786dd4daa906f20fc322a615bc5b6e2 / https://github.com/jnewbery/bitcoin/tree/pr9768.4

jnewbery committed:
Use logging in individual tests

jnewbery commented:
@laanwj @MarcoFalke I think this is a real improvement and helps debugging and fixing test cases when they fail. I've addressed https://github.com/bitcoin/bitcoin/pull/9768#issuecomment-280268516 in #9780 which builds off this PR. Any chance I can convince you to take a look at this?

ryanofsky commented on qa/rpc-tests/proxy_test.py:
```diff
@@ -83,13 +83,13 @@ def setup_nodes(self):
         # Note: proxies are not used to connect to local nodes
         # this is because the proxy to use is based on CService.GetNetwork(), which return NET_UNROUTABLE for localhost
         args = [
-            ['-listen', '-debug=net', '-debug=proxy', '-proxy=%s:%i' % (self.conf1.addr),'-proxyrandomize=1'], 
```
Line 72 `Warning: testing without local IPv6 support` is now a print again, not sure if this was intentional.

jnewbery replied:
yes - proxy_test does a lot of work in the `__init__()` method, which is called before the temp directory is set up. Logging requires the temp directory, so we can't use `self.log` until we're in `main()` and `_start_logging()` has been called.

I left this as a `print()` rather than trying to do anything clever to coerce the test into being log friendly.

ryanofsky reviewed (commented):
utACK f546d06ac21e2953c0dc0b1d8fef459b186781ce

Thanks for supporting `-l0`.

> I think this is a real improvement and helps debugging and fixing test cases when they fail.

Definitely agree. It should be especially useful to easily see interleaved python and bitcoin debug.log output while debugging.

MarcoFalke commented:
Concept ACK. Going to give some feedback later tonight.


On Wed, Mar 8, 2017 at 3:42 PM, John Newbery <notifications@github.com> wrote:
> @laanwj @MarcoFalke I think this is a real improvement and helps debugging
> and fixing test cases when they fail. I've addressed #9768 (comment) in
> #9780 which builds off this PR. Any chance I can convince you to take a look
> at this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

MarcoFalke commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -182,12 +181,38 @@ def main(self):
         else:
-            print("Failed")
+            self.log.error("Test failed. Test logging available at %s/test_framework.log", self.options.tmpdir)
             sys.exit(1)
```
You might need to shutdown (flush) the logs here.

jnewbery replied:
Done.

MarcoFalke commented on qa/rpc-tests/p2p-segwit.py:
```diff
@@ -908,7 +907,7 @@ def test_witness_tx_relay_before_segwit_activation(self):
             self.test_node.announce_tx_and_wait_for_getdata(tx, timeout=2)
-            print("Error: duplicate tx getdata!")
+            self.log.error("Error: duplicate tx getdata!")
             assert(False)
```
You don't need the assert(False) when you pass any AssertionError below. Though maybe a fix for an unrelated follow up pr.

jnewbery replied:
That's pre-existing behaviour, which is unrelated to this PR. I agree it should be fixed in a different PR.

MarcoFalke commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -182,12 +181,38 @@ def main(self):
+        # Format logs the same as bitcoind's debug.log with microprecision (so log files can be concatenated and sorted)
+        formatter = logging.Formatter(fmt = '%(asctime)s.%(msecs)03d000 %(name)s (%(levelname)s): %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
+        fh.setFormatter(formatter)
+        ch.setFormatter(formatter)
```
I think the print to stdout can use a formatter that produces smaller lines. (Maybe without the timestamp?) But no strong  opinion.

jnewbery replied:
I specifically set the format of the logs to be the same as what bitcoind produces with microsecond logging.

maflcko reviewed (commented):
utACK f546d06

jnewbery commented on qa/rpc-tests/invalidateblock.py:
```diff
@@ -18,51 +18,49 @@ def __init__(self):
-        for i in range(3):
-            print(i,self.nodes[i].getblockcount())
         assert(self.nodes[2].getblockcount() == 3)
         assert(self.nodes[0].getblockcount() == 4)
```
Done.

jnewbery commented on qa/rpc-tests/test_framework/mininode.py:
```diff
@@ -1742,17 +1737,14 @@ def got_data(self):
         if self.state != "connected" and not pushbuf:
             raise IOError('Not connected, no pushbuf')
-        self.show_debug_msg("Send %s" % repr(message))
+            logger.debug("Send message to %s:%d: %s" % (self.dstaddr, self.dstport, repr(message)))
```
Oops. Fixed.

jnewbery commented on qa/rpc-tests/test_framework/test_framework.py:
```diff
@@ -182,12 +181,38 @@ def main(self):
+        if self.options.trace_rpc:
+            rpc_logger = logging.getLogger("BitcoinRPC")
+            rpc_logger.setLevel(logging.DEBUG)
+            rpc_handler = logging.StreamHandler()
```
No reason. Just an oversight. Now fixed.

jnewbery commented:
Thanks for the feedback @MarcoFalke . I believe the latest commit should address all your comments.

MarcoFalke commented:
Ok, will take a quick look later and plan to merge this before the meeting.

On Thu, Mar 9, 2017 at 3:37 PM, John Newbery <notifications@github.com>
wrote:

> Thanks for the feedback @MarcoFalke <https://github.com/MarcoFalke> . I
> believe the latest commit should address all your comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9768#issuecomment-285368237>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmvz4ztjuB7cndYVx9sVdeT2RIPdMiks5rkA6mgaJpZM4MCCBG>
> .
>

Sjors commented:
This works if you call `test/functional/bumpfee.py --loglevel=INFO`, but not if you call `test/functional/test_runner.py --loglevel=INFO bumpfee` (which just shows dots). Not sure if that's intended, but it did [confuse me](https://botbot.me/freenode/bitcoin-core-dev/2017-11-27/?msg=93962703&page=2).

jnewbery commented:
documentation is here: https://github.com/bitcoin/bitcoin/blob/master/test/README.md#test-logging :

> when run through the test_runner harness, all logs are written to test_framework.log and no logs are output to the console.

Feel free to open a PR if you think the documentation could be improved.

Sjors commented:
@jnewberry it's probably sufficient if `/test_runner.py --help` says that next to the `--loglevel` flag (or if it doesn't show that option).

jnewbery commented:
The help text for `--loglevel` comes from `test_framework.py`. `test_runner.py` grabs that help text and appends it to its own help text. I suppose that test_runner could remove the `--loglevel` help before displaying it. The relevant code is here: https://github.com/bitcoin/bitcoin/blob/a89221873a3ee2451c73b41bbe2d99d36f439d31/test/functional/test_runner.py#L258

Sjors commented:
Thanks, I'll take a look.
